packages/langium/src/parser/langium-parser.ts

```diff
@@ -403,6 +423,123 @@ export class LangiumParserErrorMessageProvider implements IParserErrorMessagePro
 
 }
 
+export interface CompletionParserResult {
+    tokens: IToken[]
+    elementStack: AbstractElement[]
+    token?: IToken
+    tokenIndex: number
+}
+
+export class LangiumCompletionParser extends AbstractLangiumParser {
+    private mainRule!: RuleResult;
+    private elementStack: AbstractElement[] = [];
+    private lastElementStack: AbstractElement[] = [];
+    private lastToken?: IToken;
+    private nextTokenIndex = 0;
+    private tokens: IToken[] = [];
+    private stackSize = 0;
+
+    action(): void {
+        // NOOP
+    }
+
+    construct(): unknown {
+        // NOOP
+        return undefined;
+    }
+
+    parse(input: string): CompletionParserResult {
+        this.resetState();
+        const tokens = this.lexer.tokenize(input);
+        this.tokens = tokens.tokens;
+        this.wrapper.input = [...this.tokens];
+        this.mainRule.call(this.wrapper);
+        this.unorderedGroups.clear();
+        return {
+            tokens: this.tokens,
+            elementStack: [...this.lastElementStack],
+            token: this.lastToken,
+            tokenIndex: this.nextTokenIndex
+        };
+    }
+
+    rule(rule: ParserRule, impl: RuleImpl): RuleResult {
+        const ruleMethod = this.wrapper.DEFINE_RULE(rule.name, this.startImplementation(impl).bind(this));
+        if (rule.entry) {
+            this.mainRule = ruleMethod;
+        }
+        return ruleMethod;
+    }
+
+    private resetState(): void {
+        this.elementStack = [];
+        this.lastElementStack = [];
+        this.lastToken = undefined;
+        this.stackSize = 0;
+        this.nextTokenIndex = 0;
```

Comment:
This doesn't reset `tokens`, as well as `mainRule`, is that intentional?

Reply:
Yes, that's intentional. `mainRule` is set once during parser initialization and marks the entry rule of the grammar. `tokens` is reset anyway once we perform our lexing routine.